File: ccfa-paper-agent-backend/app/schemas/agent.py
from typing import Any, List, Literal, Optional, Union

from pydantic import BaseModel, Field


class AgentRequest(BaseModel):
    projectId: str
    threadId: str
    userMessage: str
    context: dict[str, Any]
    model: Optional[str] = None


class UpdateProjectMetaPatch(BaseModel):
    type: Literal["updateProjectMeta"]
    payload: dict[str, Any] = Field(default_factory=dict)


class ProposeFileChangePatch(BaseModel):
    type: Literal["proposeFileChange"]
    folderType: str
    fileId: str
    summary: str
    newContent: str


class AppendSystemMessagePatch(BaseModel):
    type: Literal["appendSystemMessage"]
    threadId: str
    content: str


class UpdateDraftParagraphStatusPatch(BaseModel):
    type: Literal["updateDraftParagraphStatus"]
    fileId: str
    paragraphId: str
    payload: dict[str, Any] = Field(default_factory=dict)


class UpdateIntroductionOutlinePatch(BaseModel):
    type: Literal["updateIntroductionOutline"]
    payload: dict[str, Any] = Field(default_factory=dict)


class UpdateScientificProblemMemoryPatch(BaseModel):
    type: Literal["updateScientificProblemMemory"]
    payload: dict[str, Any] = Field(default_factory=dict)


class AgentReferenceRequest(BaseModel):
    id: str
    title: str
    semanticScholarPaperId: str = ""
    year: Optional[int] = None
    venue: str = ""
    authors: List[str] = Field(default_factory=list)
    citationCount: Optional[int] = None
    paperUrl: str = ""
    pdfUrl: str
    externalIds: dict[str, Any] = Field(default_factory=dict)
    relevanceReason: str
    whyUsefulForThisProject: str
    suggestedReferenceScope: Literal["coreReferences", "optionalReferences"] = "optionalReferences"
    usefulForSections: List[str] = Field(default_factory=list)
    status: Literal["pending", "accepted", "rejected", "parsing", "added", "failed"] = "pending"


AgentPatch = Union[
    UpdateProjectMetaPatch,
    ProposeFileChangePatch,
    AppendSystemMessagePatch,
    UpdateDraftParagraphStatusPatch,
    UpdateIntroductionOutlinePatch,
    UpdateScientificProblemMemoryPatch,
]


class AgentResponse(BaseModel):
    content: str
    patches: Optional[List[AgentPatch]] = None
    referenceRequests: Optional[List[AgentReferenceRequest]] = None


class HealthResponse(BaseModel):
    status: str
    model: str
    provider: str


class SessionClearResponse(BaseModel):
    cleared: bool
    projectId: str
    threadId: Optional[str] = None
    clearedSessionCount: int = 0


class LocalConfigRequest(BaseModel):
    deepseekApiKey: Optional[str] = None
    deepseekBaseUrl: Optional[str] = None
    deepseekModel: Optional[str] = None
    semanticScholarApiKey: Optional[str] = None
    semanticScholarBaseUrl: Optional[str] = None
    mineruApiToken: Optional[str] = None
    mineruParseMode: Optional[str] = None


class LocalConfigResponse(BaseModel):
    saved: bool = True
    envPath: str
    deepseekApiKeyConfigured: bool
    deepseekBaseUrl: str
    deepseekModel: str
    semanticScholarApiKeyConfigured: bool
    semanticScholarBaseUrl: str
    mineruApiTokenConfigured: bool
    mineruParseMode: str


class LocalShutdownResponse(BaseModel):
    message: str
    backendPids: List[int] = Field(default_factory=list)
    frontendPids: List[int] = Field(default_factory=list)

File: ccfa-paper-agent-frontend/src/types/agent.ts
import type {
  DraftParagraph,
  FolderType,
  ParagraphWritingStatus,
  ReferencePaperMeta
} from "./file";
import type { ScientificProblemMemory, WritingStatus } from "./project";
import type { IntroductionOutline } from "./project";

export type AgentContext = {
  projectMeta: {
    id: string;
    paperTitle: string;
    targetVenue: string;
    writingStatus: WritingStatus;
    writingProgress: string;
    backendWorkspacePath?: string;
    updatedAt: string;
  };
  introductionOutline?: IntroductionOutline;
  scientificProblemMemory?: ScientificProblemMemory;
  folderSummary: Record<FolderType, number>;
  files: Array<{
    id: string;
    name: string;
    folderType: FolderType;
    mimeType: string;
    size: number;
    uploadedAt: string;
    localPath?: string;
    hasPendingChange: boolean;
    parseStatus?: string;
    hasMarkdownContent: boolean;
    hasImageData: boolean;
  }>;
  referencePaperMetas: Array<{
    fileId: string;
    fileName: string;
    folderType: FolderType;
    meta: ReferencePaperMeta;
  }>;
  imageAssets: Array<{
    id: string;
    name: string;
    caption?: string;
    markdownReference: string;
  }>;
  draftManuscripts: Array<{
    id: string;
    name: string;
    folderType: "draftManuscripts";
    localPath?: string;
    contentText?: string;
    draftParagraphs?: DraftParagraph[];
  }>;
  referencePapers: Array<{
    id: string;
    name: string;
    folderType: "coreReferences" | "optionalReferences";
    localPath?: string;
    contentText?: string;
    referenceMeta?: ReferencePaperMeta;
  }>;
};

export type AgentResponse = {
  content: string;
  patches?: AgentPatch[];
  referenceRequests?: AgentReferenceRequest[];
};

export type AgentReferenceRequest = {
  id: string;
  title: string;
  semanticScholarPaperId?: string;
  year?: number;
  venue?: string;
  authors: string[];
  citationCount?: number;
  paperUrl?: string;
  pdfUrl: string;
  externalIds?: Record<string, unknown>;
  relevanceReason: string;
  whyUsefulForThisProject: string;
  suggestedReferenceScope: "coreReferences" | "optionalReferences";
  usefulForSections?: string[];
  status: "pending" | "accepted" | "rejected" | "parsing" | "added" | "failed";
  errorMessage?: string;
};

export type AgentProgressEvent = {
  type:
    | "thinking"
    | "tool_start"
    | "tool_end"
    | "retrieving"
    | "writing"
    | "done"
    | "error";
  message: string;
  createdAt: string;
  data?: Record<string, unknown>;
};

export type AgentStreamEvent =
  | {
      type: "progress";
      event: AgentProgressEvent;
    }
  | {
      type: "final";
      response: AgentResponse;
    }
  | {
      type: "error";
      message: string;
    };

export type AgentPatch =
  | {
      type: "updateProjectMeta";
      payload: Partial<{
        paperTitle: string;
        targetVenue: string;
        writingStatus: WritingStatus;
        writingProgress: string;
      }>;
    }
  | {
      type: "proposeFileChange";
      folderType: FolderType;
      fileId: string;
      summary: string;
      newContent: string;
    }
  | {
      type: "appendSystemMessage";
      threadId: string;
      content: string;
    }
  | {
      type: "updateDraftParagraphStatus";
      fileId: string;
      paragraphId: string;
      payload: Partial<{
        writingStatus: ParagraphWritingStatus;
        userAssignedHeading: string;
      }>;
    }
  | {
      type: "updateIntroductionOutline";
      payload: Partial<IntroductionOutline> & {
        paragraphs: IntroductionOutline["paragraphs"];
      };
    }
  | {
      type: "updateScientificProblemMemory";
      payload: Partial<ScientificProblemMemory>;
    };
